### PR TITLE
[#175848172] Integrated PM Client

### PR DESCRIPTION
**Pull request by GiovanniMancini**
**Description**
This PR proposes to integrate in `io-pay` the PM client developed for IO APP.

**Proposed Changes**
- Added the PM client and its dependencies (based on old PM spec)
- Improved test suite for `fetch utils`
- Limited but extensible test suite for PM Client
- Removed in `defaultRetryingFetch` the dependency from a custom `whatwg-fetch` module

**How to test**
I added unit tests for `fetch utils` and `PM client`. Please note that to avoid mocking fetches I used the following approach https://kentcdodds.com/blog/stop-mocking-fetch , based on an `express` app which better suits our stack as an alternative to the `msw` package

### CODE REVIEW

File: src/utils/whatwg-fetch.js [hunk 1]
```diff
@@ -0,0 +1,522 @@
+// this file is a copy of https://github.com/github/fetch/tree/v2.0.4
+// the lines marked with PATCH has been changed to make this
```
Comment (gunzip) [inline src/utils/whatwg-fetch.js:2]: do you need this to work with react-native ? if you're using it in the browser you can use the unpatched package, if you're using it in node I suggest to use node-fetch

File: src/__test__/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { debug as cdebug } from 'console';
+
+describe('index', () => {
```
Comment (gunzip) [inline src/__test__/index.test.ts:3]: try to not to commit stubs

File: src/api/__test__/pagopaApi.test.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+import pm from './pm';
+
+// Client for the PagoPA PaymentManager
+describe('Weak Test Suite', () => {
+  // eslint-disable-next-line functional/no-let
+  let pmMockServer: Server = {} as Server;
+
+  beforeAll(() => {
+    pmMockServer = pm.listen(5000, '127.0.0.1');
```
Comment (gunzip) [inline src/api/__test__/pagopaApi.test.ts:15]: it looks like these are integration tests (no unit), can we mark them as such (ie. pagopaApi.integration.test.ts) in order to tell them from unit tests?

File: src/types/pagopa.ts [hunk 4]
```diff
@@ -0,0 +1,403 @@
+import { fromNullable } from 'fp-ts/lib/Option';
```
Comment (gunzip) [inline src/types/pagopa.ts:1]: this will be hard to maintain in sync with the IO App types, probably these should belong to a common package ie. https://github.com/pagopa/io-pagopa-commons

File: src/utils/input.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import * as t from 'io-ts';
```
Comment (gunzip) [inline src/utils/input.ts:1]: ditto

**Review by gunzip — COMMENTED** [review]
it looks like there's a lot of duplication from the io-app repository, these will be hard to maintain in sync in the mid term, I suggest to put the code in some common package

File: src/utils/whatwg-fetch.js (hunk 1 above)
Comment (GiovanniMancini) [reply src/utils/whatwg-fetch.js:2]: I agree. I put similar consideration as comments in the function `defaultRetryingFetch` defined in fetch utils. Since some other custom fetches in `fetch utils` relies on it, I left the patched file, but I'm planning to remove it

File: src/__test__/index.test.ts (hunk 2 above)
Comment (GiovanniMancini) [reply src/__test__/index.test.ts:3]: Thank you for pointing out. It was just the first useless test I used to develop the pipelines

File: src/api/__test__/pagopaApi.test.ts (hunk 3 above)
Comment (GiovanniMancini) [reply src/api/__test__/pagopaApi.test.ts:15]: The test against PM test server is an integration test.. I agree. But I used the tests against the local servers only as a way of unit testing without mocking fetches. The approach is similar to the one proposed here https://kentcdodds.com/blog/stop-mocking-fetch . Maybe I should separate the two kind of tests in different files.

Comment (gunzip) [reply src/api/__test__/pagopaApi.test.ts:15]: maybe I'm wrong, but it looks like a different approach: in the article you've linked to the author mocks network responses (unit tests) using msw (that's something you can do with [nock](https://github.com/nock/nock) in NodeJS context) while here we are statring a real http server (which is slower = integration test)

Comment (GiovanniMancini) [reply src/api/__test__/pagopaApi.test.ts:15]: Obviously you are right: I'm starting a real http server, not mocking its responses. Nevertheless, in my opinion the approach is _similar_ (has the same advantages, minus the performances, plus the flexibility) to the one proposed in the reference, since handlers in the express application just return a response, as if they were mocked. In my opinion, using a mock or another _double_ for a system, like a stub or a fake object, can't change the nature of the test. So if it's a unit test when mocking server responses, it should remain a unit test when using a fake server which replies with the same responses. Obviously if I'm supposed to adhere to a stricter definition of unit testing I will change the file names as you suggested, or if you think I'd better mock the responses I could change the tests. Anyway thank you for the comment, it   stimulated me to think deeper

Comment (gunzip) [reply src/api/__test__/pagopaApi.test.ts:15]: Apart from worst performance, there are some subtle differences with the two approaches:

1. using your approach you have to configure and start a new http server for every test suite (which may run concurrently). this introduces some maintenance complexity since, while your application grows, you'll need to keep the mock server implementation on par (ie. adding middlewares for different output serialization and so on). this effort is avoidable mocking the responses directly;
1. willing or not, you're testing the implementation of the server as well. for example, your tests may fail because of some express misconfiguration or because the server port is busy. that's the reason they're actually _integration_ tests. it's important (in unit tests) to limit the reason a test may fail, while with your approach this surface is broadening

While I see some downsides, I see no real gain in using a real server for testing. Can you point out some?

Comment (GiovanniMancini) [reply src/api/__test__/pagopaApi.test.ts:15]: I understand your points, but the main reasons that pushed me to favour a real server are the following
- Test the case of network failure, or server unavailability, cases that in my opinion are of a great value for the citizen, since she should be warned. Mocking the responses assumes that _there is_ a response.
- Use the same test architecture for unit tests (or something that I call unit tests :-) ) and true integration tests, in which the `express` application served by the true http server can be the same or partially the same of the one used in backend

Comment (gunzip) [reply src/api/__test__/pagopaApi.test.ts:15]: that's right, just make this distinction clear (unit vs integration) and don't mix them since integration tests may be ran against production (so we must have a way to group them)

File: package.json [hunk 6]
```diff
@@ -35,9 +43,15 @@
     "eslint-plugin-prettier": "^3.1.4",
     "eslint-plugin-react": "^7.21.5",
     "eslint-plugin-sonarjs": "^0.5.0",
-    "italia-tslint-rules": "^1.1.3",
-    "italia-utils": "^6.2.0",
-    "jest": "^25.2.7",
+    "express": "^4.17.1",
+    "faker": "^5.1.0",
+    "fp-ts": "1.12.0",
```
Comment (infantesimone) [inline package.json:48]: I think you can use the fp-ts 1.17.4, or is there some reason to use 1.12.0?

Comment (cloudify) [reply package.json:48]: in fact we must use `1.17.4` that is the same version used by `io-ts-commons` to avoid having multiple versions (will break the build): https://github.com/pagopa/io-ts-commons/blob/master/package.json#L47

Comment (GiovanniMancini) [reply package.json:48]: No, I simply sticked to the version used in IO-APP. I'll give it a try. Thank you for pointing out

**Comment by infantesimone** [comment]
LGTM 👍

**Review by infantesimone — APPROVED** [review]

File: package.json [hunk 7]
```diff
@@ -4,13 +4,16 @@
...
+  "generate:pagopa-api-definitions": "rimraf definitions/pagopa && mkdir -p definitions/pagopa && gen-api-models --api-spec $npm_package_pagopa_api --out-dir ./definitions/pagopa --no-strict --request-types --response-decoders --default-error-type undefined",
   "scripts": {
     "generate": "npm-run-all generate:*",
+    "generate:pagopa-api-definitions": "rimraf generated/definitions/pagopa && mkdir -p generated/definitions/pagopa && gen-api-models --api-spec $npm_package_pagopa_api --out-dir ./generated/definitions/pagopa --no-strict --request-types --response-decoders --default-error-type undefined",
     "generate:pagopa-api-walletv2": "rimraf generated/definitions/pagopa/walletv2 && mkdir -p generated/definitions/pagopa/walletv2  && gen-api-models --api-spec $npm_package_pagopa_api_walletv2 --out-dir ./generated/definitions/pagopa/walletv2 --no-strict --request-types --response-decoders",
     "lint": "eslint . -c .eslintrc.js --ext .ts,.tsx",
     "lint-autofix": "eslint . -c .eslintrc.js --ext .ts,.tsx --fix",
-    "test": "jest",
-    "test:coverage": "jest --coverage",
+    "test": "jest --detectOpenHandles",
```
Comment (cloudify) [inline package.json:15]: Why is this needed?

File: src/api/__test__/pagopaApi.integration.test.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import { Millisecond } from 'italia-ts-commons/lib/units';
+import { PaymentManagerClient } from '../pagopa';
+import { defaultRetryingFetch } from '../../utils/fetch';
+import * as myFetch from '../../utils/fetch';
+
+// Client for the PagoPA PaymentManager
+describe('Weak Test Suite', () => {
```
Comment (cloudify) [inline src/api/__test__/pagopaApi.integration.test.ts:7]: I suppose we can come up with a better name here :)

File: src/types/pagopa.ts (hunk 4 above)
Comment (cloudify) [reply src/types/pagopa.ts:1]: These types are "patches" to the generated types, due to the fact that the specs are wrong - there is a parallel stream of work to fix the specs so we can have the correct types generated and we can get rid of this file (and the same in io-app)

File: src/utils/fetch.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+import { left, right } from 'fp-ts/lib/Either';
+import { fromEither, TaskEither } from 'fp-ts/lib/TaskEither';
+import { calculateExponentialBackoffInterval } from 'italia-ts-commons/lib/backoff';
+import { AbortableFetch, retriableFetch, setFetchTimeout, toFetch } from 'italia-ts-commons/lib/fetch';
+import { RetriableTask, TransientError, withRetries } from 'italia-ts-commons/lib/tasks';
+
+import { Millisecond } from 'italia-ts-commons/lib/units';
+import nodeFetch from 'node-fetch';
+
+// The old following import has been substituted by an explicit declaration
```
Comment (cloudify) [inline src/utils/fetch.ts:15]: We can get rid of this comment

File: src/utils/fetch.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+// { fetchMaxRetries, fetchTimeout } from "../config";
+
+const fetchMaxRetries = 5;
```
Comment (cloudify) [inline src/utils/fetch.ts:18]: These values should be moved to a shared config file

File: src/utils/fetch.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+//
+// Default fetch configured with a short timeout and an exponential backoff
+// retrying strategy - suitable for calling the backend APIs that are supposed
+// to respond quickly.
+
+export function defaultRetryingFetch(timeout: Millisecond = fetchTimeout, maxRetries: number = fetchMaxRetries) {
+  // Override default react-native fetch with whatwg's that supports aborting
+
+  // NOTE: In fact, react-native supports Aborting Controller. Please check
```
Comment (cloudify) [inline src/utils/fetch.ts:45]: What's the point of having this comment here? Perhaps we should open a story in the app backlog

File: src/utils/fetch.ts [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  // https://github.com/facebook/react-native/blob/5e36b0c6eb2494cefd11907673aa018831526750/RNTester/js/XHRExampleAbortController.js
+  // Despite being defined in whatwg specs, the abort controller is defined in a different package.
```
Comment (cloudify) [inline src/utils/fetch.ts:47]: I don't understand this comment

File: src/utils/fetch.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+  // eslint-disable-next-line functional/immutable-data
+  (global as any).AbortController = require('abort-controller');
+
+  // NOTE: I left the following line commented since loading the custom fetch module ./whatwg-fetch.js
+  // is the approach of the IO APP. The only benefit of the custom file is to bind the "global"
+  // variable, which is required by Typescript. The package node-fetch alreday resolves
+  // such issue, so I switched to it
+
+  // require('./whatwg-fetch');
```
Comment (cloudify) [inline src/utils/fetch.ts:58]: We can get rid of this line

File: src/utils/fetch.ts [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+  return retryingFetch((global as any).fetch, timeout, maxRetries);
+}
+
+//
+// Fetch with transient error handling. Handle error that occurs once or at unpredictable intervals.
+//
+
+//
+// Fetch with transient error handling. Handle error that occurs once or at unpredictable intervals.
```
Comment (cloudify) [inline src/utils/fetch.ts:71]: Duplicate comment

File: src/utils/fetch.ts [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+//
+export function retryLogicForTransientResponseError(
```
Comment (cloudify) [inline src/utils/fetch.ts:73]: This function can be moved to io-ts-commons, would you make a PR? Cc @balanza

File: src/utils/input.ts (hunk 5 above)
Comment (cloudify) [reply src/utils/input.ts:1]: We will refactor code between this project and the app at a later time

File: src/utils/locale.ts [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+// return the current locale set in the device (this could be different from the app supported languages)
+// export const getCurrentLocale = (): Locales => I18n.currentLocale();
+
+/**
+ * return the primary component of the current locale (i.e: it-US -> it)
+ * if the current locale (the language set in the device) is not a language supported by the app
+ * the fallback will returned
+ * @param fallback
+ */
+export const getLocalePrimaryWithFallback = (fallback: Locales = 'it') => fallback;
```
Comment (cloudify) [inline src/utils/locale.ts:12]: Not sure if this is relevant here as this was made for react native's l12n apis

File: tsconfig.json [hunk 17]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
 
-    "target": "esnext",
+    "target": "es2019",
```
Comment (cloudify) [inline tsconfig.json:4]: Perhaps we should target a less recent is standard, @vito80ba what's your opinion?

**Review by cloudify — COMMENTED** [review]
@GiovanniMancini Sorry for being late to the review, I've left some comments, you can make another PR

**Comment by GiovanniMancini** [comment]
> @GiovanniMancini Sorry for being late to the review, I've left some comments, you can make another PR
Thank you for the review and suggestions, I apologize for having merged it too early but I needed some files for subsequent work on fetches

File: src/api/__test__/pagopaApi.integration.test.ts (hunk 8 above)
Comment (GiovanniMancini) [reply src/api/__test__/pagopaApi.integration.test.ts:7]: :-) Yes of course, I'll fix it, but the tests are indeed weak since I've tested only the `getSession`. I'm planning to add tests while we implement the subsequent use cases

File: src/utils/fetch.ts (hunk 10 above)
Comment (GiovanniMancini) [reply src/utils/fetch.ts:18]: I agree.. I think this module should be refactored as a part of this story https://www.pivotaltracker.com/story/show/176339507, so I've done only the minimal interventions to decouple it from IO App and make the tests work

File: src/utils/fetch.ts (hunk 11 above)
Comment (GiovanniMancini) [reply src/utils/fetch.ts:45]: The comment explains the goal of the function `defaultRetryingFetch` in the IO App. It's an old comment, I should have removed it, as `iopay` stack is not based on `react-native`. We already have a story to add the `Abort Controller` to fetches https://www.pivotaltracker.com/story/show/176339507

File: src/utils/fetch.ts (hunk 12 above)
Comment (GiovanniMancini) [reply src/utils/fetch.ts:47]: This comment is related to the fact that the `Abort Controller` is not imported from `whatwg-fetch` but from a different package/polyfill, so after I switched to `node-fetch` I thought that `whatwhg-fetch` was no longer needed in our scenario. Moreover I though that also in a `react-native` scenario we didn't need `whatwg-fetch`, as `react-native` 's core implements both `fetch` and `Abort Controller`.

File: src/utils/fetch.ts (hunk 15 above)
Comment (GiovanniMancini) [reply src/utils/fetch.ts:73]: Yes, of course

File: src/utils/locale.ts (hunk 16 above)
Comment (GiovanniMancini) [reply src/utils/locale.ts:12]: I imported `input.ts` as it is a dependency of `types/pagopa.ts`. Maybe I could purge it from useless definitions. `getLocalePrimaryWithFallback` is used in `PaymentManagerClient`, so I left it.

File: package.json (hunk 7 above)
Comment (GiovanniMancini) [reply package.json:15]: Because if the tests fail to connect to the `http server` that is spawned, `close` method I used in `afterAll()` is not a good teardown strategy. Substantially the server is force exited. In that case `detectOpenHandles` allows a clean exit or helps to detect the problem.

**Comment by cloudify** [comment]
> @cloudify Thank you for the review and suggestions, I apologize for having merged it too early but I needed some files for subsequent work on fetches


@GiovanniMancini need to apologize, we decided to merge the PR on Monday, I was just late in reviewing the code, you did the right thing
